### DISCUSSION TITLE
PyPI name normalization

### DISCUSSION THREAD

**Discussion by olii** [renovatebot/renovate#28028] [Request Help]
I found a curious case when renovate bot couldn't find a package published on a private repository index (Artifactory).

For short, lets call this package `foo.bar`:
```
[tool.poetry]
name = "foo.bar"
...
```

I build this package and published it to the artifactory. The name of the archive was changed by the build tool (normalization), meaning that a dot was converted to an underscore.
```
$ poetry build
$ find dist/
dist/foo_bar-0.5.0.tar.gz
dist/foo_bar-0.5.0-py3-none-any.whl
```

Next, only the source package was published to the index (`tar.gz`). In other projects we have this in the `requirements.txt` file:
```
foo.bar >= 0.1.0
```

Renovate bot logs the error:
```
Failed to look up pypi package foo.bar
```

I dug deep and added some logging to discover what is wrong with the resolver and noticed the code that expects the package published in the `tar.gz` format to be named as `foo.bar` but allows `foo_bar`  when in the `wheel` format.

Here is the (simplified) content of the Artifactory index:
```
<!DOCTYPE html>
<html><head><title>Simple Index</title><meta name="api-version" value="2" /></head><body>
<a href="../../foo-bar/0.1.0/foo_bar-0.1.0.tar.gz" data-requires-python="&gt;=3.10,&lt;4.0" rel="internal">foo_bar-0.1.0.tar.gz</a>
<a href="../../foo.bar/0.2.0/foo_bar-0.2.0.tar.gz" data-requires-python="&gt;=3.10,&lt;4.0" rel="internal">foo_bar-0.2.0.tar.gz</a>
<a href="../../foo.bar/0.3.0/foo_bar-0.3.0.tar.gz" data-requires-python="&gt;=3.10,&lt;4.0" rel="internal">foo_bar-0.3.0.tar.gz</a>
<a href="../../foo.bar/0.4.0/foo_bar-0.4.0.tar.gz" data-requires-python="&gt;=3.10,&lt;4.0" rel="internal">foo_bar-0.4.0.tar.gz</a>
<a href="../../foo.bar/0.5.0/foo_bar-0.5.0.tar.gz" data-requires-python="&gt;=3.10,&lt;4.0" rel="internal">foo_bar-0.5.0.tar.gz</a>
</body></html>
```

The query [here](https://github.com/renovatebot/renovate/blob/0ba9e35fcc70494f6e3b59851eff2922ab875ab2/lib/modules/datasource/pypi/index.ts#L239) correctly identifies all links and starts parsing versions in `PypiDatasource.extractVersionFromLinkText()`. Here the issue is clear. `.tar.gz` packages are not normalized:

```
packageName == 'foo.bar'
```
```js
  private static extractVersionFromLinkText(
    text: string,
    packageName: string,
  ): string | null {
    // source packages
    const srcText = PypiDatasource.normalizeName(text);
    const srcPrefix = `${packageName}-`;
    const srcSuffix = '.tar.gz';
    if (srcText.startsWith(srcPrefix) && srcText.endsWith(srcSuffix)) {
      return srcText.replace(srcPrefix, '').replace(regEx(/\.tar\.gz$/), '');
    }
...
```

Shouldn't there be a better normalization? Two functions should be better documented:
* `normalizeDepName from './common'`,
* `PypiDatasource.normalizeName`.

`extractVersionFromLinkText()` should be improved. 

# Additonal links
* https://peps.python.org/pep-0625/#specification

What are your suggestions?

**Comment by not7cd** [comment]
I was recently refactoring this part of the codebase. I have created `normalizeDepName from './common'` so it could be reused across different modules. Furthermore, I don't really understand the role of the `PypiDatasource.normalizeName` method. I see that wheels have [their own way of name normalization, which compared to PEP 503 uses `_`](https://packaging.python.org/en/latest/specifications/binary-distribution-format/#escaping-and-unicode).

https://peps.python.org/pep-0503/#normalized-names
